lightningd/peer_htlcs.c

```diff
@@ -673,31 +747,23 @@ static bool peer_accepted_htlc(struct channel *channel,
 	}
 
 	/* Unknown realm isn't a bad onion, it's a normal failure. */
+	/* FIXME: if we want hooks to handle foreign realms we should
+	 * move this check to the hook callback. */
 	if (rs->hop_data.realm != 0) {
 		*failcode = WIRE_INVALID_REALM;
 		goto out;
 	}
 
-	if (rs->nextcase == ONION_FORWARD) {
-		struct gossip_resolve *gr = tal(ld, struct gossip_resolve);
-
-		gr->next_onion = serialize_onionpacket(gr, rs->next);
-		gr->next_channel = rs->hop_data.channel_id;
-		gr->amt_to_forward = rs->hop_data.amt_forward;
-		gr->outgoing_cltv_value = rs->hop_data.outgoing_cltv;
-		gr->hin = hin;
-
-		req = towire_gossip_get_channel_peer(tmpctx, &gr->next_channel);
-		log_debug(channel->log, "Asking gossip to resolve channel %s",
-			  type_to_string(tmpctx, struct short_channel_id,
-					 &gr->next_channel));
-		subd_req(hin, ld->gossip, req, -1, 0,
-			 channel_resolve_reply, gr);
-	} else
-		handle_localpay(hin, hin->cltv_expiry, &hin->payment_hash,
-				rs->hop_data.amt_forward,
-				rs->hop_data.outgoing_cltv);
+	/* It's time to package up all the information and call the
+	 * hook so plugins can interject if they want */
+	hook_payload = tal(hin, struct htlc_accepted_hook_payload);
+	hook_payload->route_step = tal_steal(hook_payload, rs);
+	hook_payload->ld = ld;
+	hook_payload->hin = hin;
+	hook_payload->channel = channel;
+	plugin_hook_call_htlc_accepted(ld, hook_payload, hook_payload);
 
+	/* Falling through here is ok, after all the HTLC locked */
 	*failcode = 0;
```

Comment:
This needs careful design.  For example, I'm pretty sure we want to separate hooks for terminating onions and non-terminating, since most hooks will want one or the other.  For terminating we'll also want a hook after we've entered it into the db, I think, and maybe an API to query the db on restart.

In short, someone needs to use the APIs in anger before they can be finalized.

Reply:
I don't think there is much difference, immediately failing the HTLC after it was accepted (calling `local_fail_htlc` -> `fail_in_htlc`) or passing through the hook and then calling `fail_in_htlc` from the hook callback doesn't really do much. Things would be different if we could prevent an HTLC being committted to in the first place (a real immediate failure, instead of commit-then-fail), but as it stands the only difference is that the two updates to the HTLC (committed then failed) are no longer in the same DB transaction.

Reply:
As for terminating vs non-terminating onions (i.e., I'm the intended recipient vs just an intermediate hop), I did not differentiate on purpose in order to allow things like short-circuiting a payment, i.e., I know the `payment_key` because I'm in an ensemble of load-balancing nodes so I can claim the funds directly.

Reply:
We need a real user of this htlc_accepted API; I'm not smart enough to design it without a concrete one :(

Consider a just-in-time invoicing system.  The current hook is in the wrong place: any of the checks in handle_localpay could still fail.

And I can't see how the plugin's db can stay in sync with lightningd's: a restart or crash can lead to one committed and one not.  For example, after plugin commits to its db, and says continue, we crash or simply shutdown.  Now htlcs_reconnect() fails the HTLC on restart.

Adding an external dependency like this is going to be painstaking work; I think we need to pick the smallest subset so we can make mistakes with limited consequences.

Reply:
IF you were interested in starting with a simpler case with potentially less downstream implications, I could imagine various hooks for when channels change state. For instance, a simple hook for when the funding transaction is confirmed, or if a channel is asked to closed. 

On the plugin side, something simple like an email/sms alert when a channel closes could be nice. Perhaps this meets the balance between some utility and less consequential.

Reply:
This is not meant to interfere with the built-in invoicing system, it is rather a complement to it, and is meant to be used at a much lower level. With this you can pretty much create an external invoicing system, that'll get queried for open invoices.

> Consider a just-in-time invoicing system. The current hook is in the wrong place: any of the checks in handle_localpay could still fail.

I added a small test that short-circuits a payment going through `l1 -> l2 -> l3` at the middle node `l2` and just claims the HTLC, leaving the invoice on `l3` unpaid. Notice that we don't use `handle_localpay` or `forward_htlc` at all if we get told to fail or resolve.

> And I can't see how the plugin's db can stay in sync with lightningd's: a restart or crash can lead to one committed and one not. For example, after plugin commits to its db, and says continue, we crash or simply shutdown. Now htlcs_reconnect() fails the HTLC on restart.

As for the problems that may emerge from `htlcs_reconnect`, notice that I avoided having a forward-to-this channel call on purpose, so that we cannot end up with an outgoing HTLC that cannot be associated with an incoming one. The incoming HTLC on the other hand immediately fails, or gets told the preimage if the plugin returns, hence we end up always in a case that we already handle correctly.

I think as it stands this is the simplest hook I can come up with: we get an HTLC and decide whether to continue, fail or resolve. Continue is the old, and proven, logic. Fail is about as simple as it gets, and resolve is just resolving an HTLC with a `payment_key` that did not come from an invoice, but somewhere else :wink:

Reply:
After discussing this in person with @rustyrussell I now understand the issue a bit better. We may end up imposing that the hook calls must be idempotent so that we can call them again in case we have an HTLC that was accepted and wasn't resolved or failed or forwarded.